src/GafferUI/GraphGadget.cpp

```diff
@@ -223,6 +232,230 @@ class SetPositionsAction : public Gaffer::Action
+		if( context )
+		{
+			if( const Gaffer::DependencyNode *dependencyNode = runTimeCast<const Gaffer::DependencyNode>( node ) )
+			{
+				Gaffer::Context::Scope s( context );
+				const Gaffer::BoolPlug *enabledPlug = dependencyNode->enabledPlug();
+				if( enabledPlug && !enabledPlug->getValue() )
+				{
+					if( const Gaffer::Plug *active = dependencyNode->correspondingInput( connectionStart ) )
+					{
+						if( const Gaffer::Plug *input = active->getInput() )
+						{
+							activeWalk( input, active, context, canceller, activePlugs, activeNodes, plugContextsVisited );
+						}
+					}
+
+					if( const Gaffer::Plug *enabledInput = enabledPlug->getInput() )
+					{
+						activeWalk( enabledInput, enabledPlug, context, canceller, activePlugs, activeNodes, plugContextsVisited );
+					}
+					return;
+				}
+			}
+
+			if( const Gaffer::Switch *switchNode = runTimeCast<const Gaffer::Switch>( node ) )
+			{
+
+				const Gaffer::Plug *activeOutput = nullptr;
+
+				if( const Gaffer::NameSwitch *nameSwitchNode = runTimeCast<const Gaffer::NameSwitch>( switchNode ) )
+				{
+					// The top level output plug is an NameValuePlug - track just the value part for
+					// specific traversal, since all the name inputs must be tracked anyway
+					const Gaffer::NameValuePlug *outputNameValue = runTimeCast<const Gaffer::NameValuePlug>( nameSwitchNode->outPlug() );
+					if( outputNameValue && ( connectionStart == outputNameValue || connectionStart == outputNameValue->valuePlug() ) )
+					{
+						activeOutput = outputNameValue->valuePlug();
+					}
+
+					// We are handling just the value inputs specially
+					for( auto &inSourcePlug : Gaffer::Plug::InputRange( *nameSwitchNode->inPlugs() ) )
+					{
+						const Gaffer::NameValuePlug *inputNameValue = runTimeCast<const Gaffer::NameValuePlug>( inSourcePlug.get() );
+
+						handledBySpecialCase.push_back( inputNameValue->valuePlug() );
+					}
+				}
+				else
+				{
+					if( connectionStart == switchNode->outPlug() )
+					{
+						activeOutput = connectionStart;
+					}
+
+					for( auto &inSourcePlug : Gaffer::Plug::InputRange( *switchNode->inPlugs() ) )
+					{
+						handledBySpecialCase.push_back( inSourcePlug.get() );
+					}
+				}
+
+				const Gaffer::Plug *active = nullptr;
+				if( activeOutput )
+				{
+					Gaffer::Context::Scope s( context );
+					active = switchNode->activeInPlug( activeOutput );
+				}
+
+				// Track the active input plug
+				if( active )
+				{
+					const Gaffer::Plug *input = active->getInput();
+					if( input )
+					{
+						activeWalk( input, active, context, canceller, activePlugs, activeNodes, plugContextsVisited );
+					}
+				}
+			}
+			else if( const Gaffer::ContextProcessor *contextProcessorNode = runTimeCast<const Gaffer::ContextProcessor>( node ) )
+			{
+				const Gaffer::Plug *input = contextProcessorNode->inPlug()->getInput();
+				if( input )
```

Comment:
I have now done the same here - though I guess it's not necessarily correct.  The question is ... if someone adds a new output plug in a subclass, does that output plug pull on the inPlug?  On the switch, NameSwitch adds the output outIndexPlug(), which does not pull on inPlug(), which means it's fine to skip tracing inPlug() when outIndexPlug is traversed.  I guess ContextVariables::combinedVariablesPlug is the same, so it is good to add this currently.  But in theory, someone could add a ContextVariables subclass that had another output that also pulled inPlug(), in which case this would now be wrong ... maybe we just hope no one does that?

Reply:
Addresed in ae63d29

Reply:
Hmm, good point. So it seems that we should fall back to the "visit all inputs" approach for any _output_ plug we don't recognise? I suspect the future work we have in mind will mean refactoring slightly to make `activeWalk()` plug-centric rather than connection-centric, so let's deal with it then. This is working for all the cases we know about, and we need to get it into people's hands.